File: global-hooks/framework/automation/task_quality_gate.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.8"
# dependencies = []
# ///
"""Task Quality Gate - TaskCompleted hook. Logs completion metrics and nudges
the agent to verify tests when the task description mentions test requirements.
Always exits 0 (observability + gentle nudge, never blocks)."""

import json, sys
from datetime import datetime, timezone
from pathlib import Path

TEST_KEYWORDS = {"test", "verify", "validate", "check"}


def main():
    try:
        raw = sys.stdin.read()
        data = json.loads(raw) if raw.strip() else {}
    except Exception:
        data = {}
    if not data:
        sys.exit(0)

    task_id = data.get("task_id", "unknown")
    task_subject = data.get("task_subject", "")
    task_desc = data.get("task_description", "")
    teammate = data.get("teammate_name", "")

    # Log completion to JSONL
    log_path = Path.home() / ".claude" / "data" / "task_completions.jsonl"
    log_path.parent.mkdir(parents=True, exist_ok=True)
    record = {
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "task_id": task_id,
        "task_subject": task_subject,
        "teammate_name": teammate,
    }
    try:
        with open(log_path, "a") as f:
            f.write(json.dumps(record) + "\n")
    except OSError:
        pass

    # Session completion count (all entries in the log)
    count = 0
    try:
        with open(log_path) as f:
            count = sum(1 for _ in f)
    except OSError:
        pass

    # Check for test requirements in description
    desc_lower = task_desc.lower()
    has_test_req = any(kw in desc_lower for kw in TEST_KEYWORDS)

    output = {}
    if has_test_req:
        output = {"hookSpecificOutput": {
            "hookEventName": "TaskCompleted",
            "additionalContext": (
                f"[Quality Gate] Task '{task_subject}' mentions testing requirements. "
                "Please confirm that all relevant tests pass before considering this done."
            )
        }}
    print(json.dumps(output))
    sys.exit(0)


if __name__ == "__main__":
    try:
        main()
    except Exception:
        pass

File: caf-hooks/src/hooks/task_quality_gate.rs
/// Task Quality Gate — TaskCompleted hook.
///
/// Python equivalent: global-hooks/framework/automation/task_quality_gate.py (75 LOC)
///
/// Behavior:
/// - Log task completion to ~/.claude/data/task_completions.jsonl
/// - Check task_description for test keywords
/// - If keywords found, inject quality gate reminder
/// - Always exits 0
use chrono::Utc;
use serde::Serialize;
use serde_json::Value;

use crate::io::{read_stdin_value, try_append_jsonl, write_output};
use crate::state::task_completions_path;
use crate::types::HookOutput;

const TEST_KEYWORDS: &[&str] = &["test", "verify", "validate", "check"];

#[derive(Serialize)]
struct TaskCompletionRecord<'a> {
    timestamp: String,
    task_id: &'a str,
    task_subject: &'a str,
    teammate_name: &'a str,
}

pub fn run() {
    let data: Value = read_stdin_value();

    if data.is_null() || data == Value::Object(Default::default()) {
        write_output(&serde_json::json!({}));
        return;
    }

    let task_id = data.get("task_id")
        .and_then(|v| v.as_str())
        .unwrap_or("unknown")
        .to_string();
    let task_subject = data.get("task_subject")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();
    let task_desc = data.get("task_description")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();
    let teammate = data.get("teammate_name")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();

    // Log to JSONL
    let record = TaskCompletionRecord {
        timestamp: Utc::now().to_rfc3339(),
        task_id: &task_id,
        task_subject: &task_subject,
        teammate_name: &teammate,
    };
    try_append_jsonl(&task_completions_path(), &record);

    // Check for test requirements in description
    let desc_lower = task_desc.to_lowercase();
    let has_test_req = TEST_KEYWORDS.iter().any(|kw| desc_lower.contains(kw));

    if has_test_req {
        let msg = format!(
            "[Quality Gate] Task '{}' mentions testing requirements. \
            Please confirm that all relevant tests pass before considering this done.",
            task_subject
        );
        let output = HookOutput::inject_context("TaskCompleted", msg);
        write_output(&output);
    } else {
        write_output(&serde_json::json!({}));
    }
}
